src/track/trackmetadatataglib.cpp

```diff
@@ -564,44 +590,104 @@ void writeTrackMetadataIntoTag(
     }
 
-    // Set the numeric year if available
-    const QDate yearDate(
-            TrackMetadata::parseDateTime(trackMetadata.getYear()).date());
-    if (yearDate.isValid()) {
-        pTag->setYear(yearDate.year());
+    // Specialized write functions for tags derived from Taglib::Tag might
+    // be able to write the complete string from trackMetadata.getYear()
+    // into the corresponding field. In this case parsing the year string
+    // here should be omitted.
+    if (0 == (writeMask & WRITE_TAG_OMIT_YEAR)) {
+        // Set the numeric year if available
+        const QDate yearDate(
+                TrackMetadata::parseDateTime(trackMetadata.getYear()).date());
+        if (yearDate.isValid()) {
+            pTag->setYear(yearDate.year());
+        }
+    }
+
+    // The numeric track number in TagLib::Tag does not reflect the total
+    // number of tracks! Specialized write functions for tags derived from
+    // Taglib::Tag might be able to handle both trackMetadata.getTrackNumber()
+    // and trackMetadata.getTrackTotal(). In this case parsing the track
+    // number string here is useless and should be omitted.
+    if (0 == (writeMask & WRITE_TAG_OMIT_TRACK_NUMBER)) {
+        // Set the numeric track number if available
+        TrackNumbers parsedTrackNumbers;
+        const TrackNumbers::ParseResult parseResult =
+                TrackNumbers::parseFromString(trackMetadata.getTrackNumber(), &parsedTrackNumbers);
+        if (TrackNumbers::ParseResult::VALID == parseResult) {
+            pTag->setTrack(parsedTrackNumbers.getActual());
+        }
+    }
+}
+
+void writeMP4Atom(
+        TagLib::MP4::Tag* pTag,
+        const TagLib::String& key,
+        const QString& str) {
+    if (str.isEmpty()) {
+        // Purge empty atoms
+        pTag->itemListMap().erase(key);
+    } else {
+        TagLib::StringList strList(toTagLibString(str));
+        pTag->itemListMap()[key] = std::move(strList);
     }
-    // Derived tags might be able to write the complete string
-    // from trackMetadata.getYear() into the corresponding field.
+}
 
-    bool trackNumberValid = false;
-    uint track = trackMetadata.getTrackNumber().toUInt(&trackNumberValid);
-    if (trackNumberValid && (track > 0)) {
-        pTag->setTrack(track);
+void writeAPEItem(
+        TagLib::APE::Tag* pTag,
+        const TagLib::String& key,
+        const TagLib::String& str) {
+    if (str.isEmpty()) {
+        // Purge empty items
+        pTag->removeItem(key);
+    } else {
+        const bool replace = true;
+        pTag->addValue(key, str, replace);
     }
 }
 
-template<typename T>
-inline void writeMP4Atom(TagLib::MP4::Tag* pTag, const TagLib::String& key,
-        const T& value) {
-    pTag->itemListMap()[key] = value;
+bool readXiphCommentField(
+        const TagLib::Ogg::XiphComment& tag,
+        const TagLib::String& key,
+        QString* pValue = nullptr) {
+    if (tag.fieldListMap().contains(key)) {
```

Comment:
1) The double lookup is inevitable, because the map returned by 
fieldListMap() is immutable. It is only mutable for TagLib::MP4, but I 
consider this a flaw of the TagLib API. This is also the reason why I need 
to introduce the function getItemListMap(TagLib::MP4) with a const cast some 
time ago.

2) We currently do not interpret an empty string list as a missing tag for 
any tag type. If we want to handle it this way we have should introduce some 
helper functions for all tag types. I will change this, explicit functions 
for testing if a tag actually exists are needed anyway.

On 01/05/2016 07:53 PM, Daniel Schürmann wrote:
>
> In src/track/trackmetadatataglib.cpp 
> <https://github.com/mixxxdj/mixxx/pull/835#discussion_r48880846>:
>
> >      }
> >  }
> >
> > -template<typename T>
> > -inline void writeMP4Atom(TagLib::MP4::Tag* pTag, const TagLib::String& key,
> > -        const T& value) {
> > -    pTag->itemListMap()[key] = value;
> > +bool readXiphCommentField(
> > +        const TagLib::Ogg::XiphComment& tag,
> > +        const TagLib::String& key,
> > +        QString* pValue = nullptr) {
> > +    if (tag.fieldListMap().contains(key)) {
>
> Ah, I see, you version fixes the implicit insert of an empty StringList.
> But now we have again a double map lookup and the .isEmpty() is gone.
>
> How about:
>
> |FieldListMap::ConstIterator it = tag.fieldListMap().find(key); if (it != 
> tag.fieldListMap().end() && !it.second.isEmpty()) { if (nullptr != pValue) 
> { *pValue = toQStringFirstNotEmpty(it.second); ... |
>
> Unfortunately "it.second" relies in the DO_NOT_DOCUMENT part of the taglib 
> header.
> Any other ideas?
>
> —
> Reply to this email directly or view it on GitHub 
> <https://github.com/mixxxdj/mixxx/pull/835/files#r48880846>.
>